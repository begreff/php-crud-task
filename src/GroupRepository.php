<?php

namespace src;

class GroupRepository
{
    private $db_conn;

    public function __construct($db_conn)
    {
        $this->db_conn = $db_conn;
    }

    public function create($parameters)
    {
        $sql = "INSERT INTO student_groups (project_id, group_number)
                VALUES (:project_id, :group_number)";
        try {
            $statement = $this->db_conn->prepare($sql);
            $statement->execute(array(
                'project_id' => $parameters['project_id'],
                'group_number' => $parameters['group_number'],
            ));
        } catch (\Exception $e) {
            exit($e->getMessage());
        }
    }
}
