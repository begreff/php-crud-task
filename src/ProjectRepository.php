<?php

namespace src;

class ProjectRepository
{
    private $db_conn;

    public function __construct($db_conn)
    {
        $this->db_conn = $db_conn;
    }

    public function all()
    {
        $sql = "SELECT * FROM projects";

        try {
            $stmt = $this->db_conn->prepare($sql);
            $stmt->execute();
            return $stmt->fetchAll(\PDO::FETCH_ASSOC);
        } catch (\PDOException $e) {
            exit($e->getMessage());
        }
    }

    public function create($parameters)
    {
        $sql = "INSERT INTO projects (title, num_groups, students_per_group)
                VALUES (:title, :num_groups, :students_per_group)";
        try {
            $stmt = $this->db_conn->prepare($sql);
            $stmt->execute(array(
                'title' => $parameters['title'],
                'num_groups' => $parameters['num_groups'],
                'students_per_group' => $parameters['students_per_group'],
            ));
            return $this->db_conn->lastInsertID();
        } catch (\Exception $e) {
            print_r($e->getMessage());
            return null;
        }
    }

    public function read($id)
    {
        $sql = "SELECT * FROM projects WHERE id = ?";

        try {
            $stmt = $this->db_conn->prepare($sql);
            $stmt->execute(array($id));
            return $stmt->fetch(\PDO::FETCH_ASSOC);
        } catch (\PDOException $e) {
            exit($e->getMessage());
        }
    }
}
