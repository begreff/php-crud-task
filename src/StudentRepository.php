<?php

namespace src;

class StudentRepository
{
    // database connection and table
    private $db_conn;

    // project properties
    public $firstname;
    public $lastname;
    public $group_number;

    public function __construct($db_conn)
    {
        $this->db_conn = $db_conn;
    }

    function all()
    {
        $sql = "SELECT * FROM students";
        try {
            $stmt = $this->db_conn->prepare($sql);
            $stmt->execute();
            return $stmt->fetchAll(\PDO::FETCH_ASSOC);
        } catch (\PDOException $e) {
            exit($e->getMessage());
        }
    }

    function create($firstname, $lastname, $project_id)
    {
        $sql = "INSERT INTO students (firstname, lastname, project_id) 
                VALUES (:firstname, :lastname, :project_id)";

        try {
            $stmt = $this->db_conn->prepare($sql);
            return $stmt->execute(array(
                ':firstname' => $firstname,
                ':lastname' => $lastname,
                ':project_id' => $project_id
            ));
        } catch (\PDOException $e) {
            exit($e->getMessage());
        }
    }

    function read($id)
    {
        $sql = "SELECT * FROM students WHERE id = ?";
        try {
            $stmt = $this->db_conn->prepare($sql);
            $stmt->execute(array($id));
            return $stmt->fetch(\PDO::FETCH_ASSOC);
        } catch (\PDOException $e) {
            exit($e->getMessage());
        }
    }

    function update($id, $group_number)
    {
        $sql = "UPDATE students 
                SET group_number = :group_number
                WHERE id = :id";
        try {
            $statement = $this->db_conn->prepare($sql);
            return $statement->execute(array(
                ':id' => $id,
                ':group_number' => $group_number
            ));
        } catch (\PDOException $e) {
            exit($e->getMessage());
        }
    }

    function delete($id)
    {
        $sql = "DELETE FROM students WHERE id = ?";
        try {
            $stmt = $this->db_conn->prepare($sql);
            return $stmt->execute(array($id));
        } catch (\PDOException $e) {
            exit($e->getMessage());
        }
    }
}

