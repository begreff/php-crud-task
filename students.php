<?php

// CORS headers
header('Access-Control-Allow-Origin: *');

require 'vendor/autoload.php';
require 'config/conn.php';

use src\StudentRepository;

// Create object for accessing the student table in the database
$studentRepo = new StudentRepository($db->getConnection());

// Create an API variable to get the HTTP method dynamically
$api = $_SERVER['REQUEST_METHOD'];

// Get the ID from the url
$id = intval($_GET['id'] ?? '');
$project_id = intval($_GET['project_id'] ?? '');

// Add a new student to the database
if ($api == 'POST') {
    $firstname = $db->sanitizeInput($_POST['firstname']);
    $lastname = $db->sanitizeInput($_POST['lastname']);
    $project_id = $db->sanitizeInput($_POST['project_id']);

    if ($studentRepo->create($firstname, $lastname, $project_id)) {
        echo 'Student added successfully.';
    } else {
        echo "Failed to create student.";
    }
}

// Delete student from the database
if ($api == 'DELETE') {
    if ($id != 0) {
        if ($studentRepo->delete($id)) {
            echo 'Student deleted successfully.';
        } else {
            echo 'Failed to delete student.';
        }
    } else {
        echo 'Student not found.';
    }
}
