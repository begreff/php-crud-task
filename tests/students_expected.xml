<?xml version="1.0" ?>
<dataset>
    <projects id="1" title="Test" num_groups="2" students_per_group="3" />
    <student_groups project_id="1" group_number="2" />
    <students id="1" firstname="thomas" lastname="perry" project_id="1" />
    <students id="2" firstname="anna" lastname="loh" project_id="1" />
</dataset>
